utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,458 @@
+    /// let trie = Char16Trie::new(ZeroVec::from_slice(trie_data.as_slice()));
+    ///
+    /// let mut iter = trie.iter();
+    /// let res = iter.next('a');
+    /// assert_eq!(res, TrieResult::Intermediate(1));
+    /// let res = iter.next('b');
+    /// assert_eq!(res, TrieResult::FinalValue(100));
+    /// let res = iter.next('c');
+    /// assert_eq!(res, TrieResult::NoMatch);
+    /// ```
+    pub fn next(&mut self, c: char) -> TrieResult {
+        if (c as u32) <= 0xffff {
+            self.next_u16(c as u16)
+        } else {
+            match self.next_u16(u16_lead(c as i32)) {
+                TrieResult::NoValue | TrieResult::Intermediate(_) => {
+                    self.next_u16(u16_tail(c as i32))
+                }
+                _ => TrieResult::NoMatch,
+            }
+        }
+    }
+
+    /// Traverses the trie from the current state for this input char.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_char16trie::char16trie::{Char16Trie, TrieResult};
+    /// use zerovec::ZeroVec;
+    ///
+    /// // A Char16Trie containing the ASCII characters 'a' and 'b'.
+    /// let trie_data = vec![48, 97, 176, 98, 32868];
+    /// let trie = Char16Trie::new(ZeroVec::from_slice(trie_data.as_slice()));
+    ///
+    /// let mut iter = trie.iter();
+    /// let res = iter.next('a');
+    /// assert_eq!(res, TrieResult::Intermediate(1));
+    /// let res = iter.next('b');
+    /// assert_eq!(res, TrieResult::FinalValue(100));
+    /// let res = iter.next('c');
+    /// assert_eq!(res, TrieResult::NoMatch);
+    /// ```
+    pub fn next_u32(&mut self, c: u32) -> TrieResult {
+        if c <= 0xffff {
+            self.next_u16(c as u16)
+        } else {
+            match self.next_u16(u16_lead(c as i32)) {
+                TrieResult::NoValue | TrieResult::Intermediate(_) => {
+                    self.next_u16(u16_tail(c as i32))
+                }
+                _ => TrieResult::NoMatch,
+            }
+        }
+    }
+
+    /// Traverses the trie from the current state for this input char.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_char16trie::char16trie::{Char16Trie, TrieResult};
+    /// use zerovec::ZeroVec;
+    ///
+    /// // A Char16Trie containing the ASCII characters 'a' and 'b'.
+    /// let trie_data = vec![48, 97, 176, 98, 32868];
+    /// let trie = Char16Trie::new(ZeroVec::from_slice(trie_data.as_slice()));
+    ///
+    /// let mut iter = trie.iter();
+    /// let res = iter.next_u16('a' as u16);
+    /// assert_eq!(res, TrieResult::Intermediate(1));
+    /// let res = iter.next_u16('b' as u16);
+    /// assert_eq!(res, TrieResult::FinalValue(100));
+    /// let res = iter.next_u16('c' as u16);
+    /// assert_eq!(res, TrieResult::NoMatch);
+    /// ```
+    pub fn next_u16(&mut self, c: u16) -> TrieResult {
+        if self.pos.is_none() {
+            return TrieResult::NoMatch;
+        }
+        let mut pos = self.pos.unwrap();
```

Comment:
```suggestion
        let mut pos = match self.pos {
            Some(p) => p,
            None => return TrieResult::NoMatch
        };
```

Reply:
Un-rusty bystander here... the suggestion looks weird :-)
Is it faster? More idiomatic? Why is it better?

Reply:
It's better because it avoids adding an unwrap (panic) to the bytecode, which helps with code size.  (I've been fighting unwrap-related code size issues a lot this year.)  The code I proposed has identical behavior to the old code, but avoids the unwrap.

Reply:
It can also be written as

```rust
        let mut pos = if let Some(p) = self.pos {
            p
        } else {
            return TrieResult::NoMatch
        };
```